### PR TITLE
Answers file templating

### LINKED ISSUE #866: Dynamic answer file naming not working
### Discussed in https://github.com/copier-org/copier/discussions/860

<div type='discussions-op-text'>

<sup>Originally posted by **nilsdebruin** November  3, 2022</sup>
Hi all!

It is possible to define an answer file in the copier.yml like for example:

`_answers_file: .copier-answers-copier-template.yml`

It would be really nice if this could be extended with a dynamic value that you could set as a jinja variable like this:

`_answers_file: .copier-answers-copier-template-{{module_name}}.yml`

Where module_name is an answer you can give in copier.yml like:

```
module_name:
  type: str
```

This would make It possible to create multiple instances of a template in the same structure like:

```
root_level/{{module_name_1}}
root_level/{{module_name_2}}
```

This would create two answer files:

```
.copier-answers-copier-template-module_name_1.yml
.copier-answers-copier-template-module_name_2.yml
```

These template instances could then be updated based on these answer files.</div>

### PR DESCRIPTION
**Pull request by EthanBar**
Closes #866 

This adds support for templating in the `_answers_file` variable, allowing multiple answers files to be generated depending on answers given in `copier.yml`.

Was unsure whether to put tests in a new file or `tests/test_answersfile.py`. Went ahead and made a new file, but let me know if I should put them elsewhere.

### CODE REVIEW

File: tests/test_answersfile_templating.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+@pytest.fixture(scope="module")
+def template_path(tmp_path_factory) -> str:
+    root = tmp_path_factory.mktemp("template")
+    build_file_tree(
+        {
+            root
+            / ".copier-answers-{{module_name}}.yml": """\
+                # Changes here will be overwritten by Copier
+                [[ _copier_answers|to_nice_yaml ]]
```
Comment (yajo) [inline tests/test_answersfile_templating.py:18]: Bug with old template markers.

```suggestion
                {{ _copier_answers|to_nice_yaml }}
```

File: tests/test_answersfile_templating.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+from .helpers import build_file_tree
+
+
+@pytest.fixture(scope="module")
+def template_path(tmp_path_factory) -> str:
+    root = tmp_path_factory.mktemp("template")
+    build_file_tree(
+        {
+            root
+            / ".copier-answers-{{module_name}}.yml": """\
```
Comment (yajo) [inline tests/test_answersfile_templating.py:16]: Follow [the docs](https://copier.readthedocs.io/en/stable/configuring/#the-copier-answersyml-file):

```suggestion
            / "{{ _copier_conf.answers_file }}.jinja": """\
```

File: tests/test_answersfile_templating.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+                # Changes here will be overwritten by Copier
+                [[ _copier_answers|to_nice_yaml ]]
+                """,
+            root
+            / "copier.yml": """\
+                _answers_file: .answers-file-changed-in-template.yml
```
Comment (yajo) [inline tests/test_answersfile_templating.py:22]: The real place where you want to template the name is here:

```suggestion
                _answers_file: .copier-answers-{{module_name}}.yml
```

File: tests/test_answersfile_templating.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        template_path,
+        tmp_path,
+        {"module_name": "mymodule"},
+        answers_file=answers_file,
+        defaults=True,
+        overwrite=True,
+    )
+    answers_file = ".copier-answers-mymodule.yml"
+    path = Path(tmp_path) / (answers_file)
+    assert path.exists()
```
Comment (yajo) [inline tests/test_answersfile_templating.py:45]: Please assert the contents too. You'll find plenty of examples across tests.

File: tests/test_answersfile_templating.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    copier.copy(
+        template_path,
+        tmp_path,
+        {"module_name": "anothermodule"},
+        defaults=True,
+        overwrite=True,
+    )
+    answers_file = ".copier-answers-anothermodule.yml"
+    path = Path(tmp_path) / (answers_file)
+    assert path.exists()
```
Comment (yajo) [inline tests/test_answersfile_templating.py:55]: Also here.

File: tests/test_answersfile_templating.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    answers_file = ".copier-answers-mymodule.yml"
+    path = Path(tmp_path) / (answers_file)
+    assert path.exists()
```
Comment (yajo) [inline tests/test_answersfile_templating.py:58]: And here.

**Review by yajo — CHANGES_REQUESTED** [review]
Hi there!

Thanks for the patch, it seems correct. Also thank you for the test, although it is buggy. We should test the expected and supported usage only.

I can see you got a bit mislead in where should you put the template and where not. Could you please also update the docs to help others using this feature?

**Comment by stefan-caraiman** [comment]
bump, is this PR abandoned? 😢

**Comment by icj217** [comment]
I opened a new PR [here](https://github.com/copier-org/copier/pull/1027) due to the lack of activity on this PR and the fact that it didn't seem to actually work as expected during my testing. @yajo Would appreciate your 👀 on that other PR.

**Comment by sisp** [comment]
@yajo This PR can be closed as it has been superseded by #1027.

**Pull request closed without merging**